File: 2/2.hs

import Debug.Trace

evenfib upto = filter even $ fib upto

fib upto = takeWhile (< upto) f
  where f = 0 : 1 : zipWith (+) f (tail f)

--fib upto = takeWhile (< upto) $ f 0 1
--  where f x y = [y + x] ++ f y (x+y) 

--main = print $ sum $ evenfib 4000000
main = trace "testing trace" $ sum $ evenfib 4000000

File: 2/2.cpp
#include <iostream>


int main() {
  int upto = 4000000;
  int last = 1;
  int snd = 0;
  int sum = 0;
  int tmp = 0;

  while (last < upto) {
    tmp = last;
    last = last + snd;
    snd = tmp;

    if (last % 2 == 0) {
      sum += last;
    } 
  }

  std::cout << sum << std::endl;
}